### PR TITLE
Fix osu star rating calculation

### PR DESCRIPTION
**Pull request by thomastanck**
The main bug was that the beatmap was not being processed prior to
having its Skill values calculated, causing stacking to be ignored in
difficulty calculation. The fix involves processing the beatmap with
OsuBeatmapProcessor.

Another minor bug was that sliders were not taking into account the
stacked position midway through the slider (PositionAt does not return
stacked position.), so I corrected by adding StackOffset.

This fixes difficulty calculation on this map: https://osu.ppy.sh/b/1531041
osu!stable suffers from the same bug, rating this map at 7* when in reality it is a 5* map. This was verified by VINXIS by creating an equivalent map with fully manual stacking.

This is the .osu file. https://pastebin.com/raw/XChF5emy

The section causing intense difficulty spike if stacking is not accounted is

```
281,266,47907,5,12,1:2:0:0:
141,175,47996,1,0,0:0:0:0:
281,266,48086,1,0,0:0:0:0:
144,178,48175,1,0,0:0:0:0:
281,266,48266,1,8,1:2:0:0:
...
281,266,53116,1,0,0:0:0:0:
244,282,53205,1,0,0:0:0:0:
281,266,53295,1,8,1:2:0:0:
248,286,53385,1,0,0:0:0:0:
281,266,53475,5,8,1:2:0:0:
```

### CODE REVIEW

File: osu.Game/Beatmaps/DifficultyCalculator.cs [hunk 1]
```diff
@@ -19,7 +19,7 @@ public abstract class DifficultyCalculator
 
     public abstract class DifficultyCalculator<T> : DifficultyCalculator where T : HitObject
     {
-        protected readonly Beatmap<T> Beatmap;
```
Comment (Aergwyn) [inline osu.Game/Beatmaps/DifficultyCalculator.cs]: Why was this removed?

Comment (thomastanck) [reply osu.Game/Beatmaps/DifficultyCalculator.cs]: Probably a mistake.

**Comment by thomastanck** [comment]
Btw, a similar bug occurs in osu!stable, probably due to the same cause (not preprocessing the beatmap, ignoring stacking). Please verify using the .osu file (https://pastebin.com/raw/XChF5emy). Use this .osz if you want the song https://osu.ppy.sh/s/720129

You can confirm that the bug exists if setting `StackLeniency: 0` doesn't change the star rating.

StackLeniency: 0 --> 7+*
StackLeniency: 0.7 --> 5+*

File: osu.Game.Rulesets.Osu/OsuDifficulty/Preprocessing/OsuDifficultyHitObject.cs [hunk 2]
```diff
@@ -95,7 +95,7 @@ private void computeSliderCursorPosition(Slider slider)
             var computeVertex = new Action<double>(t =>
             {
                 // ReSharper disable once PossibleInvalidOperationException (bugged in current r# version)
-                var diff = slider.PositionAt(t) - slider.LazyEndPosition.Value;
+                var diff = slider.PositionAt(t) + slider.StackOffset - slider.LazyEndPosition.Value;
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Osu/OsuDifficulty/Preprocessing/OsuDifficultyHitObject.cs]: How about we expose `StackedPositionAt(double time)` from `Slider`? We'll probably want to use this elsewhere.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Review by smoogipoo — APPROVED** [review]
lgtm

**Pull request merged**